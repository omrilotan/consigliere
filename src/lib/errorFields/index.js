export const ERROR_FIELDS = [

  // Error native getters
  'message',
  'stack',
  'code',

  // inherited (not own property)
  'name',

  // non standard browser fields
  'fileName',
  'lineNumber',
  'columnNumber',

  // SystemError
  'address',
  'dest',
  'errno',
  'info',
  'path',
  'port',
  'syscall',

  // OpenSSL error properties
  'opensslErrorStack',
  'function',
  'library',
  'reason',

  // custom
  'details',
  'description'
]
