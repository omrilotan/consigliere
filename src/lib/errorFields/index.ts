/** A list of common error getters and properties */
export const ERROR_FIELDS: readonly string[] = Object.freeze([
  // Error native getters
  "message",
  "stack",
  "code",

  // inherited (not own property)
  "name",

  // non standard browser fields
  "fileName",
  "lineNumber",
  "columnNumber",

  // SystemError
  "address",
  "dest",
  "errno",
  "info",
  "path",
  "port",
  "syscall",

  // OpenSSL error properties
  "opensslErrorStack",
  "function",
  "library",
  "reason",

  // custom
  "details",
  "description",
]);
