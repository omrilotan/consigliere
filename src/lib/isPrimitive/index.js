export const isPrimitive = (value) =>
  value === null || !["function", "object"].includes(typeof value);
