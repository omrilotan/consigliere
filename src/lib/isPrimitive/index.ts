/** Check wheather a value is primitive */
export const isPrimitive = (value: any): boolean =>
  value === null || !["function", "object"].includes(typeof value);
