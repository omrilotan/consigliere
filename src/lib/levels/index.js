/**
 * @type {string[]}
 * @desc Ordered list of all levels.
 */
export const LEVELS = Object.freeze(
  ['trace', 'debug', 'verbose', 'info', 'warn', 'error', 'critical']
)
