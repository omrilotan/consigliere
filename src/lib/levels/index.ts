export type DefaultLevels =
  | "trace"
  | "debug"
  | "verbose"
  | "info"
  | "warn"
  | "error"
  | "critical";

/** Ordered list of all levels. */
export const LEVELS: readonly DefaultLevels[] = Object.freeze([
  "trace",
  "debug",
  "verbose",
  "info",
  "warn",
  "error",
  "critical",
]);
