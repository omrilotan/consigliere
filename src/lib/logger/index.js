import { log } from "../log/index.js";
import { LEVELS } from "../levels/index.js";

/**
 * @param {string[]} [levels=LEVELS]
 * @param {string} [level]
 * @param {function} [device=console.log]
 */
export class Logger {
  #levels = LEVELS;
  #minimal = 0;
  #device = console.log;
  #parser = "json";
  constructor({
    levels = LEVELS,
    level = levels.at(0),
    device = console.log,
    parser = "json",
  } = {}) {
    this.#setDevice(device);
    this.#setLevels(levels);
    this.#setLevel(level);
    this.#parser = parser;

    return new Proxy(this, {
      get(logger, prop) {
        switch (prop) {
          case "toString":
            return () => `Logger(${levels[logger.#minimal]})`;
          case "level":
            return logger.#levels[logger.#minimal];
          case "levels":
            return logger.#levels;
          case "device":
            return logger.#device;
          default:
            if (!levels.includes(prop)) {
              throw new RangeError(
                `Logger level must be one of [${levels.join(
                  ","
                )}]. Instead got [${prop}].`
              );
            }
            return levels.indexOf(prop) < logger.#minimal
              ? () => undefined
              : log.bind({ level: prop, device, parser: logger.#parser });
        }
      },
      set(logger, prop, value) {
        if (Object.isFrozen(logger)) {
          return logger;
        }

        switch (prop) {
          case "level":
            logger.#setLevel(value);
            break;
          case "levels":
            logger.#setLevels(value);
            break;
          case "device":
            logger.#setDevice(value);
            break;
          default:
            throw new Error(
              "Setting properties is only allowed for [level,levels,device]"
            );
        }
        return logger;
      },
    });
  }

  #setLevel(level) {
    const index = this.#levels.indexOf(level);
    if (index === -1) {
      throw new RangeError(
        `level must be one of [${this.#levels.join(
          ","
        )}]. Instead got [${level}].`
      );
    }
    this.#minimal = index;
  }

  #setLevels(levels) {
    if (!Array.isArray(levels)) {
      throw new TypeError(
        `levels must be an array, instead got ${typeof levels} (${levels})`
      );
    }

    if (levels.some((level) => typeof level !== "string")) {
      throw new TypeError(
        `levels must be an array of strings, instead got ${levels.map(
          (level) => typeof level
        )} (${levels.join(", ")})`
      );
    }

    this.#levels = levels;
  }

  #setDevice(device) {
    if (typeof device !== "function") {
      throw new TypeError(
        `device must be a function, instead got ${typeof device} (${device})`
      );
    }
    this.#device = device;
  }
}
