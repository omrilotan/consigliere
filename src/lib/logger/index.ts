import { log } from "../log";
import { NORMALISE } from "../parsers";
import { LEVELS } from "../levels";

/**
 * Logger class. Can be used to create multiple loggers with different settings.
 */
export class Logger {
  #levels: string[] | readonly string[] = LEVELS;
  #minimal: number = 0;
  #device: Function = console.log;
  #parser: false | Function = NORMALISE;
  #fields: Record<string, any>;
  #dynamicFields: () => Record<string, any>;
  [key: string]: any;
  constructor({
    levels = LEVELS,
    level = levels.at(0),
    device = console.log,
    parser = NORMALISE,
    fields = {},
    dynamicFields = () => ({}),
  }: {
    levels?: string[] | readonly string[];
    level?: string;
    device?: (message?: any, ...optionalParams: any[]) => void;
    parser?: (input: any) => string;
    fields?: Record<string, any>;
    dynamicFields?: () => Record<string, any>;
  } = {}) {
    this.#setDevice(device);
    this.#setLevels(levels);
    this.#setLevel(level);
    this.#setParser(parser);
    this.#setFields(fields);
    this.#setDynamicFields(dynamicFields);

    return new Proxy(this, {
      get(logger: Logger, prop: string | symbol) {
        if (typeof prop !== "string") {
          return;
        }
        switch (prop) {
          case "toString":
            return () => `Logger(${levels[logger.#minimal]})`;
          case "level":
            return logger.#levels[logger.#minimal];
          case "levels":
            return logger.#levels;
          case "device":
            return logger.#device;
          default:
            if (!levels.includes(prop)) {
              throw new RangeError(
                `Logger level must be one of [${levels.join(
                  ",",
                )}]. Instead got [${prop}].`,
              );
            }
            return levels.indexOf(prop) < logger.#minimal
              ? () => Promise.resolve(undefined)
              : log.bind({
                  level: prop,
                  device,
                  parser: logger.#parser,
                  fields: logger.#fields,
                  dynamicFields: logger.#dynamicFields,
                });
        }
      },
      set(logger: Logger, prop: string | symbol, value): boolean {
        if (Object.isFrozen(logger)) {
          return true;
        }

        switch (prop) {
          case "level":
            logger.#setLevel(value);
            break;
          case "levels":
            logger.#setLevels(value);
            break;
          case "device":
            logger.#setDevice(value);
            break;
          default:
            throw new Error(
              "Setting properties is only allowed for [level,levels,device]",
            );
        }
        return true;
      },
    });
  }

  #setLevel(level: string): void {
    const index = this.#levels.indexOf(level);
    if (index === -1) {
      throw new RangeError(
        `level must be one of [${this.#levels.join(
          ",",
        )}]. Instead got [${level}].`,
      );
    }
    this.#minimal = index;
  }

  #setLevels(levels: string[] | readonly string[]): void {
    if (!Array.isArray(levels)) {
      throw new TypeError(
        `levels must be an array, instead got ${typeof levels} (${levels})`,
      );
    }

    if (levels.some((level) => typeof level !== "string")) {
      throw new TypeError(
        `levels must be an array of strings, instead got ${levels.map(
          (level) => typeof level,
        )} (${levels.join(", ")})`,
      );
    }

    this.#levels = levels;
  }

  #setParser(parser: false | Function): void {
    if (parser !== false && typeof parser !== "function") {
      throw new TypeError(
        `parser must be a function or the boolean "false", instead got ${typeof parser} (${parser})`,
      );
    }
    this.#parser = parser;
  }

  #setDevice(device: Function): void {
    if (typeof device !== "function") {
      throw new TypeError(
        `device must be a function, instead got ${typeof device} (${device})`,
      );
    }
    this.#device = device;
  }

  #setFields(fields: object): void {
    if (fields.toString() !== "[object Object]") {
      throw new TypeError(
        `fields must be an object, instead got ${typeof fields} (${fields})`,
      );
    }
    this.#fields = fields;
  }

  #setDynamicFields(dynamicFields: () => Record<string, any>): void {
    if (typeof dynamicFields !== "function") {
      throw new TypeError(
        `dynamicFields must be a function, instead got ${typeof dynamicFields} (${dynamicFields})`,
      );
    }
    this.#dynamicFields = dynamicFields;
  }
}
