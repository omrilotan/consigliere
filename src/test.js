/* eslint-env mocha */

import { jest } from "@jest/globals";
import { Logger } from "./lib/logger";

let logger;
let mdl;

const lastLog = () => JSON.parse(console.log.mock.calls.at(-1).at(0));

describe("logger", () => {
  beforeAll(async () => {
    jest.spyOn(console, "log");
    mdl = await import(`./index.js?${Date.now()}`);
    ({ logger } = mdl);
  });
  beforeEach(() => jest.resetAllMocks());
  afterAll(() => jest.clearAllMocks());
  it("exposes full interface", () => {
    expect(Object.keys(mdl)).toMatchSnapshot();
  });
  it("pass all levels for default logger", () => {
    expect(console.log).toHaveBeenCalledTimes(0);
    logger.info("Hello", { key: "Value" });
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(lastLog().level).toBe("info");
    logger.debug("Hello", { key: "Value" });
    expect(console.log).toHaveBeenCalledTimes(2);
    expect(lastLog().level).toBe("debug");
    logger.info("Hello", { key: "Value" });
    expect(console.log).toHaveBeenCalledTimes(3);
    expect(lastLog().level).toBe("info");
    logger.warn("Hello", { key: "Value" });
    expect(console.log).toHaveBeenCalledTimes(4);
    expect(lastLog().level).toBe("warn");
    logger.error("Hello", { key: "Value" });
    expect(console.log).toHaveBeenCalledTimes(5);
    expect(lastLog().level).toBe("error");
  });
});
