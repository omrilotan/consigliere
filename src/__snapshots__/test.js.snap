// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger exposes full interface 1`] = `
Array [
  "Logger",
  "NORMALISE",
  "NORMALISE_VALUES",
  "RAW",
  "logger",
]
`;
